jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/WebSocketOverHTTP2Test.java

```diff
@@ -391,23 +397,70 @@ public void testServerTimeout() throws Exception
         assertThat(serverEndpoint.closeCode, equalTo(StatusCode.SHUTDOWN));
         assertThat(serverEndpoint.closeReason, containsStringIgnoringCase("timeout"));
         assertNotNull(serverEndpoint.error);
 
         // Wait for timeout on client.
-        assertTrue(clientEndpoint.closeLatch.await(timeout * 2, TimeUnit.MILLISECONDS));
+        assertTrue(clientEndpoint.closeLatch.await(serverIdleTimeout * 2, TimeUnit.MILLISECONDS));
         assertThat(clientEndpoint.closeCode, equalTo(StatusCode.SHUTDOWN));
         assertThat(clientEndpoint.closeReason, containsStringIgnoringCase("timeout"));
         assertNull(clientEndpoint.error);
     }
 
+    @Test
+    public void testHTTP2DisabledFallbackToHTTP1() throws Exception
+    {
+        startServer(container -> container.addMapping("/echo", (rq, rs, cb) ->
+        {
+            assertThat(rq.getConnectionMetaData().getHttpVersion(), equalTo(HttpVersion.HTTP_1_1));
+            return new EchoSocket();
+        }));
+        AbstractHTTP2ServerConnectionFactory h2c = connector.getBean(AbstractHTTP2ServerConnectionFactory.class);
+        h2c.setConnectProtocolEnabled(false);
+
+        startClient(clientConnector -> List.of(new ClientConnectionFactoryOverHTTP2.HTTP2(new HTTP2Client(clientConnector)), HttpClientConnectionFactory.HTTP11));
+
+        EventSocket clientEndpoint = new EventSocket();
+        URI uri = URI.create("ws://localhost:" + connector.getLocalPort() + "/echo");
+        Session session = wsClient.connect(clientEndpoint, uri).get(5, TimeUnit.SECONDS);
+
+        String text = "websocket";
+        session.sendText(text, Callback.NOOP);
+
+        String message = clientEndpoint.textMessages.poll(5, TimeUnit.SECONDS);
+        assertNotNull(message);
+        assertEquals(text, message);
+
+        session.close(StatusCode.NORMAL, null, Callback.NOOP);
+        assertTrue(clientEndpoint.closeLatch.await(5, TimeUnit.SECONDS));
+        assertEquals(StatusCode.NORMAL, clientEndpoint.closeCode);
+    }
+
+    @Test
+    public void testHTTP2DisabledButForced() throws Exception
+    {
+        startServer(container -> container.addMapping("/echo", (rq, rs, cb) -> new EchoSocket()));
+        AbstractHTTP2ServerConnectionFactory h2c = connector.getBean(AbstractHTTP2ServerConnectionFactory.class);
+        h2c.setConnectProtocolEnabled(false);
+
+        startClient(clientConnector -> List.of(new ClientConnectionFactoryOverHTTP2.HTTP2(new HTTP2Client(clientConnector)), HttpClientConnectionFactory.HTTP11));
+
+        EventSocket clientEndpoint = new EventSocket();
+        URI uri = URI.create("ws://localhost:" + connector.getLocalPort() + "/echo");
+        ClientUpgradeRequest upgradeRequest = new ClientUpgradeRequest(uri);
+        // Force WebSocket upgrade with HTTP/2.
+        upgradeRequest.setHttpVersion(HttpVersion.HTTP_2.asString());
+
+        assertThrows(ExecutionException.class, () -> wsClient.connect(clientEndpoint, upgradeRequest).get(5, TimeUnit.SECONDS));
```

Comment:
Could you please add an assertion checking the type of the `ExecutionException`'s cause?

Reply:
Done.